### ISSUE TITLE
Add a repl

### ISSUE THREAD

**Issue by brson** [dada-lang/dada#78]
If dada is to be capable of interactive use then it needs to be architected and tested early. And some people (like me!) can usefully dogfood even a basic repl.

I think we should add a repl and get it under test in a way that it can maintain parity with the batch interpreter, probably put it on the playground.

I have never implemented a repl before, but to my mind the main thing that distinguishes a repl is that it maintains a persistent environment of local variables that can be referenced by future input lines, something like an endless stack frame. This is something that has always been hard for Rust, with a lot of efforts at repls simply growing a large buffer of statements that are continually reexecuted. It'll be the hard part of a dada repl too, but that's why we need to figure it out now.

To get started I think we can quickly achieve a few things:

- All the basic repl UI features of line reading, signal handling, history management, via rustyline
- Multiline parsing by scanning for open token trees
- Distinguishing between items and expressions
- Compilation of items by:
    - maintaining a map of item names to source code
    - each evaluation, recreating a single source text containing all the items and updating the db.
 - Compilation of expressions by:
     - Wrapping them in a function and interpreting that function

It's going to take a lot of experimentation to figure out how to deal with expressions.

I think the repl will need to distinguish between variable assignment expressions and non-variable-assignment expressions, but even non-variable-assignment expressions will have to reference previous bindings.

@nikomatsakis has suggested a pattern of code generation that should work well to get started with expressions that reference variables, while leaning hard on dada's incremental compilation:

Given this sequence of inputs:

```
> x = 22
> y = x + 22
```

Transform it into:

```
async fn step1() {
    x = 22
    step2(x).await
}

async fn step2(x) {
    y = x + 22
    step3(x, y).await
}

async fn step3(x, y) {
    y
}
```

For every step the generated function simply passes on every variable to a stub function for the next step, preserving all the variable bindings for the next step's use. When the next step evaluates it replaces its stub function with its own definition and creates a stub for the next-next step.

dada's incremental compilation should make this quite efficient without much effort.

There are plenty of problems to figure out with this pattern, but it's worth starting and seeing how far it can go.

For testing, we'll probably add a mode that actually starts another process and pipes stdin/out. We should be able to make it so all batch tests that compile and run successfully can be fed into the repl and behave the same, using the same reference files, though confirming that both modes issue the same e.g. warning diagnostics could be tricky. It may also be easy use the same reference files for files that compile successfully but have runtime errors. Compiler errors could be harder, not sure.

**Comment by brson** [comment]
I have a branch where I am working on this and have the UI, items, and basic expressions working, https://github.com/brson/dada/tree/repl, but the code is a mess while I am learning how the compiler works. I could submit a cleaned up wip pr for review in a week or so, but I'm enjoying tinkering on it for now and don't know when I'll actually want to merge something.

**Comment by brson** [comment]
One problem with the previously-referenced pattern for handling functions and expressions is that it doesn't allow for persistence of the program state between repl steps, at least that I can see. It doesn't recompile the whole program, but it does reevaluate the whole program, every step.

**Comment by nikomatsakis** [comment]
I think you were right to raise this question early, as the design wasn't really made with a REPL in mind.

**Comment by brson** [comment]
> I do think we could separate out the state of the interpreter from the `Interpreter` itself -- the main thing would be that it would be bogus to modify an existing function's source code. That said, we could also consider moving the interpreter state to be managed by salsa, which would permit jumping back and forth in time etc, as well as gracefully handling incremental recompilation, but I am skeptical. It doesn't seem like the right fit to me -- but maybe if we think a bit about the queries, it's plausible.

A repl could arrange never to modify existing items by quietly giving them all unique names, though then it has to figure out how to rewrite every expression to use the most recent uniquely-named item.

Having the interpreter be managed by salsa is an interesting prospect that would make it behave in ways that are unique and interesting, though with a presumably huge complexity/performance penalty, and one that may impose on aot-compiled semantics in unforseen ways. It's a fun idea, but I agree it's probably not the way to go.



> The other option yeah would be to have the interpreter state be its own value. I made some steps in this direction by threading the `Interpreter` as an argument everywhere instead of storing it in the `StackFrame` values. I think the state of the interpreter itself is just a `Vec<StackFrame>`, and if we know that we are only appending new functions, the existing stack frames ought not to be disturbed by changes, though really we're abusing salsa there because you're not _supposed_ to assume that (e.g.) entity-ids and the like persist when you mutate the inputs.

As I see it now the repl needs to maintain something like a single ever-changing stack frame, but I don't think it looks like the current dada StackFrame since e.g. there isn't a single BIR or that expresses a function that is slowly growing forever. At the moment it seems fine to me to wrap every repl expression in a function and just store a bunch of values in the repl "stack frame", and munge any spans that come out of the compiler to look more like they looked when the user typed them.

@nikomatsakis Assuming the repl never replaces any items, is it ok to just hold on to interpreter values outside of the interpreter, and pass them into a new interpreter every time the repl runs a new expr?

**Comment by brson** [comment]
How a rewind/replay capability might help an interpreter is something I haven't thought about.

**Comment by brson** [comment]
The main reason I don't love reexecuting the entire history every step, both in a repl, and during rewinding, is because it replays side-effects. With dada having a runtime in control of I/O it could be possible to have it record the side effects and selectively lie to the program about I/O so as not to e.g. re-print to the terminal.

**Comment by nikomatsakis** [comment]
@brson yes, at the moment there are no notable side-effects, but this is partly why I made a point of creating a `Kernel` abstraction that is meant to encapsulate all side-effects, so that we can capture inputs and replay them etc.

**Comment by nikomatsakis** [comment]
I wonder @brson is what we want is to add a new kind of "thing" to the IR -- not a function but more like a top-level "expression". Then we could add new ones. 

(I mean, tbh, it *might* even make sense if we didn't have a main function but just let people dump code in the outside context, although that's always bugged me for some reason.)

**Comment by nikomatsakis** [comment]
@brson ok, I've been thinking about a refactoring to the interpreter that seems relevant here. I'm planning to restructure it so that there is a core data structure (the "machine") that just tracks:

* the heap
* the data on the stack, along with current PC

and then there is a separate bit of code that takes a machine and moves execution forward by one step. I think I will move away from the idea that a call in dada corresponds to a call in Rust.

This is stemming from the work on permissions. For various reasons, I think it'd be better to separate out the "data being operated on" into its own data structure and out of the Rust heap/stack. It will be less efficient, but I think flexibility and analyzability are way more important to optimize for. (I expect to have a compiler eventually that is for performance.)

**Comment by brson** [comment]
The new interpreter structure gives the repl more possibilities, and I've made some progress, able to set and reference variables. I'll post a wip pr soon so you can track it.

The approach I'm using now is similar to what you originally suggested, where every repl step I create a stub function to contain the next step, and pass the entire environment into it. When the stub is called I save the machine state for later - then on the next step replace the top stack frame with the new function while preserving all the argument slots from the replaced frame.

**Comment by brson** [comment]
Unassigning myself. Busy with other things lately.

**Comment by brson** [comment]
Stale.
